project/_vars.sls

```diff
@@ -1,7 +1,8 @@
 {% set root_dir = "/var/www/" + pillar['project_name'] + "/" %}
 
-{% macro get_primary_ip(ifaces) -%}
-  {{ ifaces.get(salt['pillar.get']('primary_iface', 'eth0'), {}).get('inet', [{}])[0].get('address') }}
+{% macro get_primary_ip(host, ifaces) -%}
+  {# use `primary_iface` from pillar, then fall back to getting the iface from the default route #}
+  {{- ifaces[salt['pillar.get']('primary_iface', salt['mine.get'](host, 'network.default_route')[host][0]['interface'])].get('inet', [{}])[0].get('address') -}}
```

Comment:
Drat, now I'm finding this unreadable.  I wonder if we could break it back down into more easily digested expressions using macros?  E.g. maybe a macro to get a host's primary interface, another to get a host interface's primary ip, and then finally a macro that uses those two to get a hosts primary ip?  Or whatever makes sense.